### DISCUSSION TITLE
Brainstorming

### DISCUSSION THREAD

**Discussion by Banana71** [probonopd/MiniDexed#407] [Ideas]
- FX:
_Reverb, Compressor, Delay and/or EQ as Master FX or even per TG_
https://github.com/probonopd/MiniDexed/pull/795

- LAN/WLAN:
_Access to the SD card via LAN/WLAN_
https://github.com/probonopd/MiniDexed/pull/783

- Keyboard Zone setting:
_Zone settings should also be adjustable directly from the UI._
`
NoteLimitLow1=0
NoteLimitHigh1=127
`

- Audio recorder:
_A nice tool would be an audio recorder that could write directly to the micro SD card. An additional switch via GPIO that can start and stop recording directly._

--------------------------------------------------
has already been done: 👍 

- Performance change via MIDI ProgramChange:
_Is it possible to change the performance via MIDI program change?_
In minidexed.ini `PerformanceSelectChannel` determines if MIDI Program Change (PC) and Bank Selection (CC BANKSEL) messages should select individual voices/banks within a tone generator or act at the system level and select performances. This mirrors the similarly named control on a DX7II. Setting this to 0 (the default) means PC messages select voices/banks on any tone generators set to the receiving MIDI channel. Setting this to a MIDI channel (1 to 16) means PC messages on that channel will select performances. Setting this to a value >16 selects "Omni" mode so any PC messages received select performances. Note when in performance mode, no PC messages will be passed on to tone generators. Also note that MIDIRXProgramChange has to be set to 1 for this to have any affect.
---------------------------------------------------
Ideas that cannot be implemented in this way and were therefore discarded: 👎 

- Cutoff:
_The filter is not part of the DX7, but controlling it via an envelope, LFO, modulation wheel or aftertouch would take the sound to new dimensions.
Is it possible to pick up the envelopes of the EG's, PEG or LFO and route them to the filter?_

There is only one filter per instance, so a filter envelope is not possible as hoped. However, the filter can be influenced externally via MIDI CC or directly in the menu.

- Pan:
_A powerful tool of the Yamaha SY99 is its own envelope with loop function for the panorama.
Is it possible to pick up the envelopes of the EG's, LFO or PEG and route them to the panorama?
Or is it possible to implement an auto pan effect?_

The panorama can only be set per TG (instance) and not "per voice per TG". But a MIDI effect "Auto Pan" would be possible.

----------------------------------------------------

What ideas do you have?

**Comment by probonopd** [comment]
About using the envelope, LFO, modulation wheel, etc., wouldn't that mean that our patches ("voice data") would be incompatible with DX7 type instruments?

**Reply by Banana71** [reply]
The DX7II has its own panorama envelope in performance mode. 
The functions would be set up in "performance mode" and would not change anything in the actual engine

![Screenshot_20221224-113443](https://user-images.githubusercontent.com/104296205/209432215-6f2df578-5de3-46ef-9c3f-321af14200a0.png)

![Screenshot_20221224-114241](https://user-images.githubusercontent.com/104296205/209432430-59d35e82-fb3f-41e8-affe-cb9981c36192.png)

**Reply by probonopd** [reply]
That's good to know. Thanks for the hint!

**Reply by pik33** [reply]
> About using the envelope, LFO, modulation wheel, etc., wouldn't that mean that our patches ("voice data") would be incompatible with DX7 type instruments?

I think no, if the additional functions are placed after existing synthesizer machine, like the external effect processor added to the output of each TG, receiving MIDI commands in parallel to the main synthesizer. 

As if the real DX7 was connected to the external effect processor. 

This effect processor can dynamically filter and pan (and doing whatever else) the audio received from a TG using assigned midi controllers and its own envelope generators.

**Comment by BobanSpasic** [comment]
Just a quick glossary:
*CED - single voice format, one parameter = one byte
*MEM - 32 voices format, compressed (more parameters combined in one byte)

VCED - DX7 (mark I) single voice data - basic data that we also use for MiniDexed. VoiceData in performance.ini is stored as VCED. SYX-files are in VMEM format.
PCED - TX7 additional data (addition to VCED) - some of the parameters can also be found in MiniDexed Performance. Also known as Function Data. In TX7 SYX-files are 32 voices as VMEM, followed by 64 Function Data blocks as PMEM. Just the first 32 Function Data blocks are used.
ACED - DX7II (mark II) additional data (addition to VCED) - some of the parameters can also be found in MiniDexed Performance. Also known as Supplement Data. In DX7II SYX-files are 32 voices as VMEM, followed by 32 additional data blocks as AMEM.

For DX7, DX7II, TX7, TX 816 etc - just the VMEM part of SYX-files is read if you transfer the voices from one to the other model.

Every new parameter you want to implement in MiniDexed - define own additional data format as supplement. VCED/VMEM should remain the same for all 6 OP DX-synths. In the fact, we already have performance.ini files - it is MiniDexed's 8x Supplemental Data.

btw. next version of MiniDexed Control Center will have conversion of PCED and ACED to MiniDexed Performance (also from Bank Dumps PMEM and AMEM).

I am in a vacation at the moment, but I'll come with some whitepapers in January. We need to define a couple of things at this point (binary format for Performances - something like own xCED and xMEM formats, MIDI header ID for Performances, legal issues if we go under F0 43 00 xMEM - Yamaha is the owner of ID 43 etc.).

**Comment by Banana71** [comment]
Zone settings should also be adjustable directly from the UI.

NoteLimitLow1=0
NoteLimitHigh1=127

**Comment by Banana71** [comment]
Performance change via MIDI ProgramChange
Is it possible to change the performance via MIDI program change?

I would imagine something like that:
Configurable in minidex.ini
E.g. PerformanceChange=0 (everything as before, ProgramChange does not change the performance)
PerformanceChange=2 (ProgramChange on channel 2 change the performance)

**Reply by probonopd** [reply]
Interesting. Do you know any hardware instruments which are doing this?
Is there a standard way in MIDI to a) switch voices and b) switch whole performances?

**Reply by Banana71** [reply]
The Yamaha Montag / MODX always works in performance mode. They reacts to program changes with another performance.

**Comment by probonopd** [comment]
Looked a bit into the features that the DX7 II has over the DX7. Would be nice to see them in MiniDexed.

**Comment by Banana71** [comment]
Could you create a drum module version for the weaker Raspberry PI 1 or an RP Zero? By that I mean a version with 8 TG's but with 2 voice polyphony. This is how you could play drum sounds and FX sound on a PI-Zero :-D. Or generally with all other RP versions 8 additional TG's with 2-fold polyphony (TG 9 - TG16)
I know this isn't something the TX816 can do, but it would be a great addition. !E Gray Matter also pushed boundaries.
!E Gray Matter for miniDexed :-D

**Reply by probonopd** [reply]
It should be possible to allow twice the numbers of TGs at half the polyphony if I am not mistaken, and I think it might be possible to go beyond 8 TGs at least on the beefier Raspberry Pi models (read: 4, 400), but one would need to think carefully about how to distribute the workload across the cores.

For best flexibility, maybe it should be made configurable how many TGs with how much polyphony are running on which cores. Maybe @rsta2 can help us looking into it?

**Reply by Banana71** [reply]
I didn't know the specifications of the Yamaha TX802 at all, I always thought it was just the rack version of the DX7IID. The TX802 can do a lot of what I want from the miniDexed. Performance change via MIDI. With only 16 voice polyphony it is still 8 part multitimbral. (RP 1 and RP Pi Zero)
[Yamaha TX802 Manual.pdf](https://github.com/probonopd/MiniDexed/files/10982291/Yamaha.TX802.Manual.pdf)

**Comment by DutchBigEd** [comment]
I'm always, well  most of the times ;) , looking for a sound for droning, and the best so far is still when I play my didgeridoo...
Could there be performances especially good for drones?

**Reply by Banana71** [reply]
do you have a sound example in which direction it should go or a song?

**Reply by DutchBigEd** [reply]
There is a recording I made with a drone that one of the best I ever made (in my opinion that is...)

https://soundcloud.com/user-406494421/drone_19_12

**Reply by DutchBigEd** [reply]
At modwiggler there is a thread about drones (not all to my taste though.. )

https://www.modwiggler.com/forum/viewtopic.php?t=90817

**Comment by Banana71** [comment]
I have absolutely no idea what I'm doing here. This is completely new territory that I am entering here. My naive implementation from the gut.

TG1&2 Engine 1             Keyzone C0 - E1
TG3&4 Engine 2             Keyzone F1 - B2
TG5 HaertBeat Sonor     Keyzone C2
TG8 Sonor                      Keyzone C#2
TG6&7 Falling ?              Keyzone D2 - 

[000105_Drone.zip](https://github.com/probonopd/MiniDexed/files/11178491/000105_Drone.zip)

https://user-images.githubusercontent.com/104296205/230600698-5e46e9ba-41fd-49ac-85a4-a9bd809f1a67.mp4

**Reply by DutchBigEd** [reply]
The 1st (lowest) octave sounds are (almost) great!
The attack and decay are to steep (for my taste).
So (for me) attacks and decay should be slow, while sustain should hold as long as the key is down.
Accept for the "ping" at C#, this is a great bell sound.

**Reply by Banana71** [reply]
> Accept for the "ping" at C#, this is a great bell sound.

Yes, I was almost certain that it had nothing to do with it :-D
If I have time I will adjust the envelopes

**Comment by Banana71** [comment]
Hi Ed @DutchBigEd , 
I read something about drone sounds and now I know that I know nothing.
Here is my sound bank and the second attempt. You can keep working on it.

[000106_Drone2.ini.zip](https://github.com/probonopd/MiniDexed/files/11183386/000106_Drone2.zip)
[000000_BananaDron.syx.zip](https://github.com/probonopd/MiniDexed/files/11183387/000000_BananaDron.zip)

<html>


<body>


Performance | TG | Voice | Notelimit | Desciption
-- | -- | -- | -- | --
Drone | TG1&2 | Engine | C0 – D#1 | Overtones disappear, hold
|  | TG3&4 | Engine 2 | E1 – G#1 | Variation
|  | TG5&6 | Engine 3 | A2 – C2 | Variation
|  | TG7 | CoffEngine | C#2 – A8 | Hold cords and play with Cutoff
|  | TG8 |   | C#2 – A8 | Volume 0, switch short to 1
| | | | | go to TG7 Cutoff, Cords above C#2


</body>

</html>

**Reply by DutchBigEd** [reply]
Thank you very much! Later today I will listen to your voice and performance

**Reply by probonopd** [reply]
At some point one has to wonder whether we should build in fetching voice banks/performances via the network, accessible from the menu. From configurable GitHub repositories. E.g., this way we could always get @Banana71's latest and greatest without messing around with the microSD card... just a thought. Kinda like Soundmondo but totally open source. Wdyt?

**Reply by Banana71** [reply]
> At some point one has to wonder whether we should build in fetching voice banks/performances via the network, accessible from the menu. From configurable GitHub repositories. E.g., this way we could always get @Banana71's latest and greatest without messing around with the microSD card... just a thought. Kinda like Soundmondo but totally open source. Wdyt?

That sounds very good, but if it weren't the third step before the first in this direction.
I could imagine the following route for improving the miniDexed.

**1. Save from Voice Data:** Changes to the Voice Data under Voice Edit, no matter in which TG, can be saved in banks. The memory location (0 - 32) can be freely selected and it is possible to overwrite patches. It would be worth considering whether to create extra RAM banks for this or whether every existing bank could be overwritten. I would define extra RAM banks and all other banks I would set a quasi MEMEROY PROTECT.

**2. Transmit voice data:**
  2.1 Transmit Voice Data for all 32 Voices: A bank can be selected on the miniDexed, which can be sent to an external device via the MIDI Out.
  2.2 Transmit Single Voice Data:
  2.3 Transmit All Voices from a Performance: All sounds from the loaded Performance are sent as a bank (32 patches) to an external device via the MIDI Out. The sounds 1 - 8 are taken from the performance and the patches 9 - 32 are filled with e.g. "Empty Voice".

**3. Loading and saving of sounds/performance over network**

That would be my thoughts on it.

**Reply by probonopd** [reply]
With "network" I actually meant "the Internet"...

**Comment by Gustavomurta** [comment]
Hi Banana71, 
About the audio recorder, would it be for example to capture data from the I2S interface (DAC Audio) and record a wave file on the sd card? I think It might be possible to do this with a Raspberry Pico.
